File: app/src/main/java/com/clearkeep/chat/signal_store/InMemorySenderKeyStore.java
package com.clearkeep.chat.signal_store;

import com.clearkeep.utilities.storage.Storage;

import org.whispersystems.libsignal.groups.SenderKeyName;
import org.whispersystems.libsignal.groups.state.SenderKeyRecord;
import org.whispersystems.libsignal.groups.state.SenderKeyStore;

import java.io.IOException;
import java.util.HashMap;
import java.util.Map;

public class InMemorySenderKeyStore implements SenderKeyStore {

  private final Map<SenderKeyName, SenderKeyRecord> store = new HashMap<>();

  private final Storage storage;

  public InMemorySenderKeyStore(Storage storage) {
    this.storage = storage;
  }

  @Override
  public void storeSenderKey(SenderKeyName senderKeyName, SenderKeyRecord record) {
    store.put(senderKeyName, record);
  }

  @Override
  public SenderKeyRecord loadSenderKey(SenderKeyName senderKeyName) {
    try {
      SenderKeyRecord record = store.get(senderKeyName);

      if (record == null) {
        return new SenderKeyRecord();
      } else {
        return new SenderKeyRecord(record.serialize());
      }
    } catch (IOException e) {
      throw new AssertionError(e);
    }
  }
}

File: app/src/main/java/com/clearkeep/chat/signal_store/InMemorySenderKeyStore.kt
package com.clearkeep.chat.signal_store

import com.clearkeep.db.signal.SignalKeyDAO
import com.clearkeep.db.signal.model.SignalSenderKey
import org.whispersystems.libsignal.groups.SenderKeyName
import org.whispersystems.libsignal.groups.state.SenderKeyRecord
import org.whispersystems.libsignal.groups.state.SenderKeyStore
import java.io.IOException
import java.util.*

class InMemorySenderKeyStore(private val signalKeyDAO: SignalKeyDAO) : SenderKeyStore {

    private val store: MutableMap<SenderKeyName, SenderKeyRecord> = HashMap()

    override fun storeSenderKey(senderKeyName: SenderKeyName, record: SenderKeyRecord) {
        store[senderKeyName] = record
        signalKeyDAO.insert(SignalSenderKey(
                senderKeyName.groupId,
                senderKeyName.sender.name,
                senderKeyName.sender.deviceId,
                record.serialize()
        ))
    }

    override fun loadSenderKey(senderKeyName: SenderKeyName): SenderKeyRecord {
        return try {
            var record = store[senderKeyName]

            if (record == null) {
                val senderKey = signalKeyDAO.getSignalSenderKey(senderKeyName.groupId, senderKeyName.sender.name, senderKeyName.sender.deviceId)
                if (senderKey != null) {
                    record = SenderKeyRecord(senderKey.senderKey)
                    // update cache
                    store[senderKeyName] = record
                }
            }

            if (record == null) {
                SenderKeyRecord()
            } else {
                SenderKeyRecord(record.serialize())
            }
        } catch (e: IOException) {
            throw AssertionError(e)
        }
    }
}